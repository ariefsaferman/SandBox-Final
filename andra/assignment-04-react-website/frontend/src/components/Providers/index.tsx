import React from "react";
import { CookiesProvider } from "react-cookie";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { store } from "../../store";

type Props = {
  children: React.ReactNode;
};

const Providers = (props: Props) => {
  return (
    <CookiesProvider>
      <Provider store={store}>
        <BrowserRouter>{props.children}</BrowserRouter>
      </Provider>
    </CookiesProvider>
  );
};

export default Providers;
