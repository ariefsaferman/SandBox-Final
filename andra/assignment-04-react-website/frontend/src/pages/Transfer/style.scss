.transfer__title {
  font-weight: 700;
  font-size: 2.5rem;
  margin-bottom: 3.75rem;
}

.vh-90 {
  height: 90vh;
}
