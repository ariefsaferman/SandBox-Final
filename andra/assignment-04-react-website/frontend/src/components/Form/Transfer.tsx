import { useEffect, useState } from "react";
import { useCookies } from "react-cookie";
import { useForm } from "react-hook-form";
import { useDispatch, useSelector } from "react-redux";
import { toast, ToastContainer } from "react-toastify";
import ITransaction from "../../interfaces/transaction";
import { RootState } from "../../store";
import { fetchUser, UserDispatch } from "../../store/slices/userSlice";
import TransactionModal from "../TransactionModal";
import "./style.scss";

type Payload = {
  amount: number;
  to: number;
  description: string;
};

const TransferForm = () => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<Payload>();
  const { user } = useSelector((state: RootState) => state.user);
  const [showModal, setShowModal] = useState(false);
  const [transaction, setTransaction] = useState<ITransaction>();
  const [cookies] = useCookies(["token"]);
  const dispatch: UserDispatch = useDispatch();
  const api = process.env.REACT_APP_API_URL;

  useEffect(() => {
    dispatch(fetchUser(cookies.token));
  }, [cookies.token, dispatch]);

  const onSubmit = (data: Payload) => {
    const requestOptions = {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: "Bearer " + cookies.token,
      },
      body: JSON.stringify(data),
    };

    fetch(api + "/transactions/transfer", requestOptions)
      .then(async (res) => {
        const resJson = await res.json();
        if (!res.ok) throw new Error(resJson.message);
        return resJson;
      })
      .then((res) => {
        if (res.data) {
          dispatch(fetchUser(cookies.token));
          setTransaction(res.data);
          setShowModal(true);
        }
      })
      .catch((err) => toast.error(err.message));
  };

  return (
    <form
      data-testid="transfer-form"
      className="form__transfer"
      onSubmit={handleSubmit(onSubmit)}
    >
      <TransactionModal show={showModal} type="Transfer" data={transaction} />
      <ToastContainer />
      <div className="form-group">
        <label htmlFor="from">From</label>
        <input
          data-testid="from-input"
          className="form-control"
          disabled
          type="number"
          value={user?.wallet_id}
        />
      </div>
      <div className="form-group">
        <label htmlFor="">To</label>
        <input
          className="form-control"
          type="number"
          placeholder="To"
          {...register("to", {
            required: true,
            setValueAs: (value) => parseInt(value),
          })}
          style={{ borderColor: errors.to && "red" }}
        />
        {errors.to && (
          <span data-testid="error" className="error">
            Please insert transfer target
          </span>
        )}
      </div>
      <div className="form-group">
        <label htmlFor="">Amount</label>
        <input
          className="form-control form-control-lg"
          type="number"
          placeholder="Amount"
          min={1000}
          max={50000000}
          {...register("amount", {
            required: {
              value: true,
              message: "Please insert amount of transfer",
            },
            min: {
              value: 1000,
              message: "Minimum transfer amount is IDR 1.000",
            },
            max: {
              value: 50000000,
              message: "Maximum transfer amount is IDR 50.000.000",
            },
            setValueAs: (value) => parseInt(value),
          })}
          style={{ borderColor: errors.amount && "red" }}
        />
        {errors.amount && (
          <span data-testid="error" className="error">
            {errors.amount?.message}
          </span>
        )}
      </div>
      <div className="form-group">
        <label htmlFor="">Description</label>
        <textarea
          className="form-control"
          placeholder="Description"
          {...register("description", { required: true })}
          style={{ borderColor: errors.description && "red" }}
        />
        {errors.description && (
          <span data-testid="error" className="error">
            Please insert description
          </span>
        )}
      </div>
      <button
        data-testid="btn-submit"
        type="submit"
        className="btn btn-lg w-100"
      >
        Send
      </button>
    </form>
  );
};

export default TransferForm;
