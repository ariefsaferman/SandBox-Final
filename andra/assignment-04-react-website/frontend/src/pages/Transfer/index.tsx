import React from "react";
import TransferForm from "../../components/Form/Transfer";
import "./style.scss";

const Transfer = () => {
  return (
    <div className="container">
      <div className="d-flex vh-90 flex-column align-items-center justify-content-center">
        <h2 className="transfer__title">Transfer</h2>
        <TransferForm />
      </div>{" "}
    </div>
  );
};

export default Transfer;
