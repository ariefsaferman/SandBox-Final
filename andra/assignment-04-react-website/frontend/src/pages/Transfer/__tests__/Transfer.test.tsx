import { fireEvent, render, screen } from "@testing-library/react";
import Transfer from "..";
import Providers from "../../../components/Providers";

test("render", () => {
  const view = render(<Transfer />, { wrapper: Providers });
  expect(view).toMatchSnapshot();
});

test("transfer form rendered", () => {
  render(<Transfer />, { wrapper: Providers });

  const form = screen.getByTestId("transfer-form");
  expect(form).toBeInTheDocument();
});

test("from input pre-filled with user wallet id", () => {
  render(<Transfer />, { wrapper: Providers });

  const fromInput = screen.getByTestId("from-input");
  expect(fromInput).not.toHaveValue("");
});

test("error when one or more input is empty", async () => {
  render(<Transfer />, { wrapper: Providers });

  const btnSubmit = screen.getByTestId("btn-submit");
  expect(btnSubmit).toBeInTheDocument();

  fireEvent.click(btnSubmit);
  const errors = await screen.findAllByTestId("error");

  expect(errors).not.toHaveLength(0);
});

test("error when amount is less than 1000", async () => {
  render(<Transfer />, { wrapper: Providers });

  const amountInput = screen.getByPlaceholderText("Amount");
  expect(amountInput).toBeInTheDocument();

  fireEvent.change(amountInput, { target: { value: 500 } });

  const btnSubmit = screen.getByTestId("btn-submit");
  expect(btnSubmit).toBeInTheDocument();

  fireEvent.click(btnSubmit);

  const error = await screen.findByText("Minimum transfer amount is IDR 1.000");
  expect(error).toBeInTheDocument();
});

test("error when amount is more than 50000000", async () => {
  render(<Transfer />, { wrapper: Providers });

  const amountInput = screen.getByPlaceholderText("Amount");
  expect(amountInput).toBeInTheDocument();

  fireEvent.change(amountInput, { target: { value: 60000000 } });

  const btnSubmit = screen.getByTestId("btn-submit");
  expect(btnSubmit).toBeInTheDocument();

  fireEvent.click(btnSubmit);

  const error = await screen.findByText(
    "Maximum transfer amount is IDR 50.000.000"
  );
  expect(error).toBeInTheDocument();
});

test("show modal when success", async () => {
  render(<Transfer />, { wrapper: Providers });

  const toInput = screen.getByPlaceholderText("To");
  expect(toInput).toBeInTheDocument();
  const amountInput = screen.getByPlaceholderText("Amount");
  expect(amountInput).toBeInTheDocument();
  const descriptionInput = screen.getByPlaceholderText("Description");
  expect(descriptionInput).toBeInTheDocument();

  fireEvent.change(toInput, { target: { value: 2 } });
  fireEvent.change(amountInput, { target: { value: 1000 } });
  fireEvent.change(descriptionInput, { target: { value: "test" } });

  const btnSubmit = screen.getByTestId("btn-submit");
  expect(btnSubmit).toBeInTheDocument();

  fireEvent.click(btnSubmit);

  const modal = await screen.findByTestId("transaction-modal");
  expect(modal).toBeInTheDocument();
});
