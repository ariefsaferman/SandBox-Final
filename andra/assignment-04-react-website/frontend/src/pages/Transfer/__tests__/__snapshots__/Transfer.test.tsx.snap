// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="d-flex vh-100 flex-column align-items-center justify-content-center"
        >
          <h2
            class="transfer__title"
          >
            Transfer
          </h2>
          <form
            class="form__transfer"
            data-testid="transfer-form"
          >
            <button
              class="d-none"
              data-bs-target="#modal"
              data-bs-toggle="modal"
              type="button"
            />
            <div
              class="modal fade"
              data-bs-backdrop="static"
              data-bs-keyboard="false"
              data-testid="transaction-modal"
              id="modal"
              tabindex="-1"
            >
              <div
                class="modal-dialog modal-dialog-centered"
              >
                <div
                  class="modal-content"
                >
                  <div
                    class="modal-header"
                  >
                    <svg>
                      success-icon.svg
                    </svg>
                    <h5
                      class="modal-title text-center"
                    >
                      Transfer
                       Success
                    </h5>
                  </div>
                  <div
                    class="modal-body"
                  >
                    <table
                      class="table table-borderless"
                    >
                      <tbody>
                        <tr>
                          <td
                            class="align-bottom"
                          >
                            Amount
                          </td>
                          <td
                            class="transaction__amount"
                          />
                        </tr>
                        <tr>
                          <td>
                            Transaction Id
                          </td>
                          <td />
                        </tr>
                        <tr>
                          <td>
                            From
                          </td>
                          <td />
                        </tr>
                        <tr>
                          <td>
                            To
                          </td>
                          <td />
                        </tr>
                        <tr>
                          <td>
                            Description
                          </td>
                          <td />
                        </tr>
                      </tbody>
                    </table>
                  </div>
                  <div
                    class="modal-footer"
                  >
                    <button
                      class="btn"
                      type="button"
                    >
                      Print
                    </button>
                    <button
                      class="btn"
                      data-bs-dismiss="modal"
                      type="button"
                    >
                      Close
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="Toastify"
            />
            <div
              class="form-group"
            >
              <label
                for="from"
              >
                From
              </label>
              <input
                class="form-control"
                data-testid="from-input"
                disabled=""
                type="number"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for=""
              >
                To
              </label>
              <input
                class="form-control"
                name="to"
                placeholder="To"
                type="number"
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for=""
              >
                Amount
              </label>
              <input
                class="form-control form-control-lg"
                max="50000000"
                min="1000"
                name="amount"
                placeholder="Amount"
                type="number"
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for=""
              >
                Description
              </label>
              <textarea
                class="form-control"
                name="description"
                placeholder="Description"
              />
            </div>
            <button
              class="btn btn-lg w-100"
              data-testid="btn-submit"
              type="submit"
            >
              Send
            </button>
          </form>
        </div>
         
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="d-flex vh-100 flex-column align-items-center justify-content-center"
      >
        <h2
          class="transfer__title"
        >
          Transfer
        </h2>
        <form
          class="form__transfer"
          data-testid="transfer-form"
        >
          <button
            class="d-none"
            data-bs-target="#modal"
            data-bs-toggle="modal"
            type="button"
          />
          <div
            class="modal fade"
            data-bs-backdrop="static"
            data-bs-keyboard="false"
            data-testid="transaction-modal"
            id="modal"
            tabindex="-1"
          >
            <div
              class="modal-dialog modal-dialog-centered"
            >
              <div
                class="modal-content"
              >
                <div
                  class="modal-header"
                >
                  <svg>
                    success-icon.svg
                  </svg>
                  <h5
                    class="modal-title text-center"
                  >
                    Transfer
                     Success
                  </h5>
                </div>
                <div
                  class="modal-body"
                >
                  <table
                    class="table table-borderless"
                  >
                    <tbody>
                      <tr>
                        <td
                          class="align-bottom"
                        >
                          Amount
                        </td>
                        <td
                          class="transaction__amount"
                        />
                      </tr>
                      <tr>
                        <td>
                          Transaction Id
                        </td>
                        <td />
                      </tr>
                      <tr>
                        <td>
                          From
                        </td>
                        <td />
                      </tr>
                      <tr>
                        <td>
                          To
                        </td>
                        <td />
                      </tr>
                      <tr>
                        <td>
                          Description
                        </td>
                        <td />
                      </tr>
                    </tbody>
                  </table>
                </div>
                <div
                  class="modal-footer"
                >
                  <button
                    class="btn"
                    type="button"
                  >
                    Print
                  </button>
                  <button
                    class="btn"
                    data-bs-dismiss="modal"
                    type="button"
                  >
                    Close
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="Toastify"
          />
          <div
            class="form-group"
          >
            <label
              for="from"
            >
              From
            </label>
            <input
              class="form-control"
              data-testid="from-input"
              disabled=""
              type="number"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for=""
            >
              To
            </label>
            <input
              class="form-control"
              name="to"
              placeholder="To"
              type="number"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for=""
            >
              Amount
            </label>
            <input
              class="form-control form-control-lg"
              max="50000000"
              min="1000"
              name="amount"
              placeholder="Amount"
              type="number"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for=""
            >
              Description
            </label>
            <textarea
              class="form-control"
              name="description"
              placeholder="Description"
            />
          </div>
          <button
            class="btn btn-lg w-100"
            data-testid="btn-submit"
            type="submit"
          >
            Send
          </button>
        </form>
      </div>
       
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
